tools/wavetool.py

```diff
@@ -217,14 +232,15 @@ def calc_thdn(wave, fs, freq):
 
     Returns
     ----------
-    A list of THD+N values for each channel
+    A list of THD+N values for each channel, unit: dB
     """
     # skip 20ms to avoid the pulse due to filtering
     time_skip = 0.02
     samples_skip = int(time_skip * fs)
     filtered = stdnotch(wave, freq, fs)
-    filtered_cut = filtered[samples_skip:,:]
-    return 10 * np.log10(np.mean(np.power(filtered_cut, 2), axis=0))
+    filtered_cut = filtered[samples_skip:]
+    return 10 * np.log10(np.mean(np.power(filtered_cut, 2), axis=0) + eps)
```

Comment:
I think we should call this "THD+N ratio" since it was measured with sine wave input. Extracts from AES spec:

_"The test signal shall be a sine wave. The analyzer shall include the standard low-pass filter (see 5.2.5). The total
signal level at the EUT output is measured and recorded. The fundamental component of the test signal present
in the output shall be removed with the standard notch filter (see 5.2.8), and the unweighted level of the residual
shall be measured. The THD+N ratio shall be the difference between the residual level and the total signal
level, expressed in dB."

and

"The test signal level shall be a sine wave at -1 dB relative to the maximum input level (see 6.2.1). "

So, the used sine level should be subtracted from the result. E.g. -81 - (-1) dB gives -80 dB.

Reply:
> The test signal level shall be a sine wave at -1 dB relative to the maximum input level 

I think we may not use the THD+N ratio here,  because our test signal is not a standard -1 dB sine wave.  our loopback wave varies in level, so an absolute value (thdn) is better than a relative value (thdn ratio)

Reply:
The problem without ratio is that a muted / a lot attenuated test might a really good THD+N value. You need to set the limit to such that an OK performing system passes it. I would never pass a result worse than -60 dB (equivalent to 10 bits performance - our implementation is 16/24/32 bits).

Reply:
@singalsu For a good quality sine wave with -30dB level, and -80dB thdn, the THD+N ratio is on -50dB. it is hard to maintain a 0 dB volume control in our loopback route, how can we deal with this condition? maybe we should keep the calculation of thdn and thdn ration, and let user to choose what they want to use?

Reply:
With current test system (headphone out -> headset adapter mic in) I think it's not expected to get any better performance. Could the pass criteria be something like -48 dB until the CI hardware can be upgraded? Testing it is much better than nothing though small issues will be left unnoticed. The THD+N alone is not good because a broken system can get a good result.

@mengdonglin The -50 dB THD+N ratio is performance of a 80's C-cassette player. Please try to escalate the need for better audio validation hardware. We really should be able to use something like -70 to -80 dB THD+N ratio as automatic test pass/fail criteria.